File: dotnet/external-user-service/UserService/shala.api/api/tenants/TenantController.cs
using FluentValidation;
using shala.api.common;
using shala.api.domain.types;
using shala.api.services;
using shala.api.startup;

namespace shala.api;

public class TenantController
{
    private readonly ITenantService _tenantService;
    private readonly IUserService _userService;
    private readonly IRoleService _roleService;
    private readonly IUserRoleService _userRoleService;
    private readonly IValidator<TenantCreateModel> _createValidator;
    private readonly IValidator<TenantUpdateModel> _updateValidator;
    private readonly IValidator<TenantSearchFilters> _searchValidator;
    private readonly ILogger<TenantController> _logger;

    public TenantController(
                        ITenantService TenantService,
                        IUserService userService,
                        IRoleService roleService,
                        IUserRoleService userRoleService,
                        IValidator<TenantCreateModel> createValidator,
                        IValidator<TenantUpdateModel> updateValidator,
                        IValidator<TenantSearchFilters> searchValidator,
                        ILogger<TenantController> logger)
    {
        _tenantService = TenantService;
        _userService = userService;
        _roleService = roleService;
        _userRoleService = userRoleService;
        _createValidator = createValidator;
        _updateValidator = updateValidator;
        _searchValidator = searchValidator;
        _logger = logger;
    }

    public async Task<IResult> Create(HttpContext context, TenantCreateModel model)
    {
        try
        {
            var currentUser = UserAuthenticationHandler.GetCurrentUser(context);
            if (currentUser == null)
            {
                return ResponseHandler.Unauthorized("User not authenticated");
            }
            var isSystemAdmin = currentUser.Role == DefaultRoles.SystemAdmin.ToString();
            if (!isSystemAdmin)
            {
                return ResponseHandler.Forbidden("Only system admin can create tenant");
            }
            var validationResult = await _createValidator.ValidateAsync(model);
            if (!validationResult.IsValid)
            {
                var errors = validationResult.Errors.ToDictionary(
                        e => e.PropertyName,
                        e => new[] { e.ErrorMessage }
                    );
                return ResponseHandler.ValidationError("Validation error", errors);
            }

            var tenantCode = model.Code;
            if (tenantCode != null)
            {
                var existingTenantWithTenantCode = await _tenantService.GetByCodeAsync(tenantCode);
                if (existingTenantWithTenantCode != null)
                {
                    return ResponseHandler.Conflict("Tenant already exists");
                }
            }
            else
            {
                tenantCode = await Helper.GetUniqueTenantCode(this._tenantService, model.Name);
                model.Code = tenantCode;
            }
            var email = model.Email;
            if (email == null)
            {
                return ResponseHandler.BadRequest("Email is required");
            }

            var existingTenantWithEmail = await _tenantService.GetByEmailAsync(email);
            if (existingTenantWithEmail != null)
            {
                return ResponseHandler.Conflict("Tenant with this email already exists");
            }

            var user = await _userService.GetByEmailAsync(email);
            if (user == null)
            {
                user = await this._userService.CreateAsync(new UserCreateModel
                {
                    Email = email,
                    UserName = email,
                    Password = model.Password,
                    CountryCode = model.CountryCode ?? null,
                    PhoneNumber = model.PhoneNumber ?? null,
                });
            }
            if (user == null)
            {
                return ResponseHandler.InternalServerError("Tenant admin user cannot be created");
            }

            var role = await _roleService.GetByNameAsync("TenantAdmin");
            if (role == null)
            {
                return ResponseHandler.InternalServerError("Role not found");
            }

            var record = await _tenantService.CreateAsync(model);
            if (record == null)
            {
                return ResponseHandler.InternalServerError("Tenant not created");
            }

            var roleAdded = await _userRoleService.AddRoleToUserAsync(user.Id, role.Id);
            if (!roleAdded)
            {
                return ResponseHandler.InternalServerError("User role not created");
            }

            return ResponseHandler.Created("Tenant created successfully", record, $"/tenants/{record.Id}");
        }
        catch (Exception ex)
        {
            return ResponseHandler.ControllerException(ex);
        }

    }

    public async Task<IResult> GetById(HttpContext context, Guid id)
    {
        try
        {
            if (id == Guid.Empty)
            {
                return ResponseHandler.BadRequest("Invalid tenant id");
            }
            var record = await _tenantService.GetByIdAsync(id);
            return record != null ?
                ResponseHandler.Ok("Tenant retrieved successfully", record, $"/tenants/{record.Id}") :
                ResponseHandler.NotFound("Tenant not found");
        }
        catch (Exception ex)
        {
            return ResponseHandler.ControllerException(ex);
        }
    }

    public async Task<IResult> GetByCode(HttpContext context, string code)
    {
        try
        {
            if (string.IsNullOrWhiteSpace(code))
            {
                return ResponseHandler.BadRequest("Invalid tenant code");
            }
            var record = await _tenantService.GetByCodeAsync(code);
            return record != null ?
                ResponseHandler.Ok("Tenant retrieved successfully", record, $"/tenants/{record.Id}") :
                ResponseHandler.NotFound("Tenant not found");
        }
        catch (Exception ex)
        {
            return ResponseHandler.ControllerException(ex);
        }
    }

    public async Task<IResult> Update(HttpContext context, Guid id, TenantUpdateModel model)
    {
        try
        {
            var hasPermissions = this.hasPermissions(context, id);
            if (!hasPermissions)
            {
                return ResponseHandler.Forbidden("You do not have permission to update this tenant");
            }
            var validationResult = await _updateValidator.ValidateAsync(model);
            if (!validationResult.IsValid)
            {
                var errors = validationResult.Errors.ToDictionary(
                        e => e.PropertyName,
                        e => new[] { e.ErrorMessage }
                    );
                return ResponseHandler.ValidationError("Validation error", errors);
            }
            var record = await _tenantService.UpdateAsync(id, model);
            if (record == null)
            {
                return ResponseHandler.NotFound("record not found");
            }
            return ResponseHandler.Ok("Tenant updated successfully", record, $"/tenants/{record.Id}");
        }
        catch (Exception ex)
        {
            return ResponseHandler.ControllerException(ex);
        }
    }

    private bool hasPermissions(HttpContext context, Guid id)
    {
        var currentUser = UserAuthenticationHandler.GetCurrentUser(context);
        if (currentUser == null)
        {
            return false;
        }
        var currentUserTenantId = currentUser.TenantId;
        var isSystemAdmin = currentUser.Role == DefaultRoles.SystemAdmin.ToString();
        if (isSystemAdmin)
        {
            return true;
        }
        if (currentUserTenantId != id)
        {
            return false;
        }
        var isTenantAdmin = currentUser.Role == DefaultRoles.TenantAdmin.ToString();
        if (!isTenantAdmin)
        {
            return false;
        }
        return true;
    }

    public async Task<IResult> Search(HttpContext context, TenantSearchFilters filters)
    {
        try
        {
            var validationResult = await _searchValidator.ValidateAsync(filters);
            if (!validationResult.IsValid)
            {
                var errors = validationResult.Errors.ToDictionary(
                        e => e.PropertyName,
                        e => new[] { e.ErrorMessage }
                    );
                return ResponseHandler.ValidationError("Validation error", errors);
            }
            var tenants = await _tenantService.SearchAsync(filters);
            return ResponseHandler.Ok("Tenants retrieved successfully", tenants);
        }
        catch (Exception ex)
        {
            return ResponseHandler.ControllerException(ex);
        }
    }

    public async Task<IResult> Delete(HttpContext context, Guid id)
    {
        try
        {
            var hasPermissions = this.hasPermissions(context, id);
            if (!hasPermissions)
            {
                return ResponseHandler.Forbidden("You do not have permission to delete this tenant");
            }
            if (id == Guid.Empty)
            {
                return ResponseHandler.BadRequest("Invalid tenant id");
            }
            var deleted = await _tenantService.DeleteAsync(id);
            return deleted ?
                ResponseHandler.Ok("Tenant deleted successfully") :
                ResponseHandler.InternalServerError("Error deleting tenant");
        }
        catch (Exception ex)
        {
            return ResponseHandler.ControllerException(ex);
        }
    }

}

File: nodejs/external-user-service/user-service/src/api/tenant/tenants/tenant.controller.ts
import express from 'express';
import { TenantService } from '../../../services/tenant/tenant.service';
import { ResponseHandler } from '../../../common/handlers/response.handler';
import { Injector } from '../../../startup/injector';
import { TenantValidator } from './tenant.validator';
import { ApiError } from '../../../common/api.error';
import { uuid } from '../../../domain.types/miscellaneous/system.types';
import { RoleService } from '../../../services/authorization/role.service';
import { UserService } from '../../../services/users/user.service';
import { UserRoleService } from '../../../services/authorization/user.role.service';
import { DefaultRoleTypes } from '../../../domain.types/authorization/enums';
import { UserCreateModel } from '../../../domain.types/users/user.types';
import { logger } from '../../../logger/logger';
import { EmailService } from "../../../modules/communication/email/email.service";
import { EmailDetails } from "../../../modules/communication/email/email.details";
import { TenantDto, TenantCreateModel, TenantUpdateModel } from "../../../domain.types/tenant/tenant.types";
import { Helper } from '../../../common/helper';
import { BaseController } from '../../../api/base.controller';

///////////////////////////////////////////////////////////////////////////////////////

export class TenantController extends BaseController {

    //#region member variables and constructors

    _service: TenantService = Injector.Container.resolve(TenantService);

    _roleService: RoleService = Injector.Container.resolve(RoleService);

    _userService: UserService = Injector.Container.resolve(UserService);

    _userRoleService: UserRoleService = Injector.Container.resolve(UserRoleService);

    _validator: TenantValidator = new TenantValidator();

    //#endregion

    create = async (request: express.Request, response: express.Response): Promise<void> => {
        let tenant: TenantDto = null;
        try {
            const model: TenantCreateModel = await this._validator.create(request);
            if (model.TenantCode === 'default') {
                throw new ApiError(400, 'Cannot create tenant with code "default"!');
            }

            await this.authorizeOne(request);

            tenant = await this._service.create(model);
            if (tenant == null) {
                throw new ApiError(400, 'Unable to create tenant.');
            }

            const tenantCode = tenant.TenantCode;
            const adminUserName = model.UserName?.toLowerCase() ?? (tenantCode + '-admin').toLowerCase();
            let existingUser = await this._userService.getByUserName(adminUserName);
            if (existingUser) {
                throw new ApiError(400, 'Username already exists');
            }

            const adminPassword = model.Password ?? Helper.generatePassword();
            const role = await this._roleService.getByName(DefaultRoleTypes.TenantAdmin);
            const userModel: UserCreateModel = {
                PhoneCode   : tenant.PhoneCode,
                PhoneNumber : tenant.PhoneNumber,
                Email       : tenant.Email,
                FirstName   : 'Admin',
                LastName    : tenant.Name,
                TenantId    : tenant.id,
                UserName    : adminUserName,
            };

            existingUser = await this._userService.getExistingUser({
                PhoneCode   : userModel.PhoneCode,
                PhoneNumber : userModel.PhoneNumber,
                Email       : userModel.Email,
                UserName    : userModel.UserName,
                TenantId    : userModel.TenantId,
            });

            if (existingUser) {
                throw new ApiError(400, 'User already exists!');
            }

            const user = await this._userService.create(userModel);
            if (user == null) {
                throw new ApiError(400, 'Unable to create tenant admin user.');
            }
            logger.info(`Tenant admin user created successfully. UserName: ${adminUserName}`);

            const userRole = await this._userRoleService.addUserRole(user.id, role.id, tenant.id);
            if (userRole == null) {
                throw new ApiError(400, 'Unable to create tenant admin user role.');
            }
            //Send email to the admin user with username and password
            await this.sendWelcomeEmail(tenant, adminUserName, adminPassword);

            ResponseHandler.success(request, response, 'Tenant added successfully!', 201, {
                Tenant    : tenant,
                AdminUser : {
                    UserName : adminUserName
                }
            });
        } catch (error) {
            ResponseHandler.handleError(request, response, error);
        }

    };

    getById = async (request: express.Request, response: express.Response): Promise<void> => {
        try {
            const id: uuid = await this._validator.getParamUuid(request, 'id');
            const tenant = await this._service.getById(id);
            if (tenant == null) {
                throw new ApiError(404, 'Tenant not found.');
            }
            ResponseHandler.success(request, response, 'Tenant retrieved successfully!', 200, {
                Tenant : tenant,
            });
        } catch (error) {
            ResponseHandler.handleError(request, response, error);
        }
    };

    getByCode = async (request: express.Request, response: express.Response): Promise<void> => {
        try {
            const code: string = await this._validator.getParamStr(request, 'code');
            const tenant = await this._service.getByCode(code);
            if (tenant == null) {
                throw new ApiError(404, 'Tenant not found.');
            }
            ResponseHandler.success(request, response, 'Tenant retrieved successfully!', 200, {
                Tenant : tenant,
            });
        }
        catch (error) {
            ResponseHandler.handleError(request, response, error);
        }
    };

    search = async (request: express.Request, response: express.Response): Promise<void> => {
        try {
            const currentUserRoles = request.currentUser.Roles;
            const roleNames = currentUserRoles.map(r => r.Name);
            if (!roleNames.includes(DefaultRoleTypes.SystemAdmin) &&
                !roleNames.includes(DefaultRoleTypes.SystemUser)
            ) {
                throw new ApiError(403, 'Unauthorized action!');
            }
            const filters = await this._validator.search(request);
            const searchResults = await this._service.search(filters);
            const count = searchResults.Items.length;
            const message =
                count === 0 ? 'No records found!' : `Total ${count} tenant records retrieved successfully!`;
            ResponseHandler.success(request, response, message, 200, {
                TenantRecords : searchResults,
            });
        } catch (error) {
            ResponseHandler.handleError(request, response, error);
        }
    };

    update = async (request: express.Request, response: express.Response): Promise<void> => {
        try {
            const id: uuid = await this._validator.getParamUuid(request, 'id');
            const tenant = await this._service.getById(id);
            if (tenant == null) {
                throw new ApiError(404, 'Tenant not found.');
            }
            await this.authorizeOne(request, null, tenant.id);
            if (tenant.TenantCode === 'default') {
                throw new ApiError(400, 'Cannot update tenant with code "default"!');
            }
            const model: TenantUpdateModel = await this._validator.update(request);
            const updatedTenant = await this._service.update(id, model);
            if (tenant == null) {
                throw new ApiError(400, 'Unable to update tenant record!');
            }
            ResponseHandler.success(request, response, 'Tenant updated successfully!', 200, {
                Tenant : updatedTenant,
            });
        } catch (error) {
            ResponseHandler.handleError(request, response, error);
        }
    };

    delete = async (request: express.Request, response: express.Response): Promise<void> => {
        try {
            const id: uuid = await this._validator.getParamUuid(request, 'id');
            const tenant = await this._service.getById(id);
            if (tenant == null) {
                throw new ApiError(404, 'Tenant not found.');
            }
            await this.authorizeOne(request, null, tenant.id);
            if (tenant.TenantCode === 'default') {
                throw new ApiError(400, 'Cannot delete tenant with code "default"!');
            }
            const deleted = await this._service.delete(id);
            ResponseHandler.success(request, response, 'Tenant deleted successfully!', 200, {
                Deleted : deleted,
            });
        } catch (error) {
            ResponseHandler.handleError(request, response, error);
        }
    };

    promoteTenantUserAsAdmin = async (request: express.Request, response: express.Response): Promise<void> => {
        try {
            const id: uuid = await this._validator.getParamUuid(request, 'id');
            const userId: uuid = await this._validator.getParamUuid(request, 'userId');
            const tenant = await this._service.getById(id);
            if (tenant == null) {
                throw new ApiError(404, 'Tenant not found.');
            }
            await this.authorizeOne(request, null, tenant.id);
            const user = await this._service.getById(userId);
            if (user == null) {
                throw new ApiError(404, 'User not found.');
            }
            const promoted = await this._service.promoteTenantUserAsAdmin(id, userId);
            ResponseHandler.success(request, response, 'User promoted as admin to tenant successfully!', 200, {
                Promoted : promoted,
            });
        }
        catch (error) {
            ResponseHandler.handleError(request, response, error);
        }
    };

    demoteAdmin = async (request: express.Request, response: express.Response): Promise<void> => {
        try {
            const id: uuid = await this._validator.getParamUuid(request, 'id');
            const userId: uuid = await this._validator.getParamUuid(request, 'userId');
            const tenant = await this._service.getById(id);
            if (tenant == null) {
                throw new ApiError(404, 'Tenant not found.');
            }
            await this.authorizeOne(request, null, tenant.id);
            const user = await this._service.getById(userId);
            if (user == null) {
                throw new ApiError(404, 'User not found.');
            }
            const demoted = await this._service.demoteAdmin(id, userId);
            ResponseHandler.success(request, response, 'User demoted as admin from tenant successfully!', 200, {
                Demoted : demoted,
            });
        }
        catch (error) {
            ResponseHandler.handleError(request, response, error);
        }
    };

    getTenantStats = async (request: express.Request, response: express.Response): Promise<void> => {
        try {
            const id: uuid = await this._validator.getParamUuid(request, 'id');
            const tenant = await this._service.getById(id);
            if (tenant == null) {
                throw new ApiError(404, 'Tenant not found.');
            }
            await this.authorizeOne(request, null, tenant.id);
            const stats = await this._service.getTenantStats(id);
            ResponseHandler.success(request, response, 'Tenant stats retrieved successfully!', 200, {
                Stats : stats,
            });
        }
        catch (error) {
            ResponseHandler.handleError(request, response, error);
        }
    };

    getTenantAdmins = async (request: express.Request, response: express.Response): Promise<void> => {
        try {
            const id: uuid = await this._validator.getParamUuid(request, 'id');
            const tenant = await this._service.getById(id);
            if (tenant == null) {
                throw new ApiError(404, 'Tenant not found.');
            }
            await this.authorizeOne(request, null, tenant.id);
            const admins = await this._service.getTenantAdmins(id);
            ResponseHandler.success(request, response, 'Tenant admins retrieved successfully!', 200, {
                Admins : admins,
            });
        }
        catch (error) {
            ResponseHandler.handleError(request, response, error);
        }
    };

    getTenantRegularUsers = async (request: express.Request, response: express.Response): Promise<void> => {
        try {
            const id: uuid = await this._validator.getParamUuid(request, 'id');
            const tenant = await this._service.getById(id);
            if (tenant == null) {
                throw new ApiError(404, 'Tenant not found.');
            }
            await this.authorizeOne(request, null, tenant.id);
            const moderators = await this._service.getTenantRegularUsers(id);
            ResponseHandler.success(request, response, 'Tenant moderators retrieved successfully!', 200, {
                Moderators : moderators,
            });
        }
        catch (error) {
            ResponseHandler.handleError(request, response, error);
        }
    };

    private sendWelcomeEmail = async (tenant: TenantDto, adminUserName: string, adminPassword: string) => {
        try {
            const emailService = new EmailService();
            var body = await emailService.getTemplate('tenant.welcome.template.html');

            body = body.replace(/{{PLATFORM_NAME}}/g, process.env.PLATFORM_NAME);
            body = body.replace(/{{PLATFORM_WEBSITE}}/g, process.env.PLATFORM_WEBSITE);
            body = body.replace(/{{PLATFORM_SUPPORT_EMAIL}}/g, process.env.PLATFORM_SUPPORT_EMAIL);
            body = body.replace(/{{TENANT_NAME}}/g, tenant.Name);
            body = body.replace('{{TENANT_ADMIN_USER_NAME}}', adminUserName);
            body = body.replace('{{TENANT_ADMIN_PASSWORD}}', adminPassword);

            const emailDetails: EmailDetails = {
                EmailTo : tenant.Email,
                Subject : `Welcome`,
                Body    : body,
            };
            logger.info(`Username -> ${adminUserName}`);
            logger.info(`Password -> ${adminPassword}`);
            logger.info(`Sending welcome email to ${tenant.Email}`);
            logger.info(`Email details: ${JSON.stringify(emailDetails)}`);

            const sent = await emailService.sendEmail(emailDetails, false);
            if (!sent) {
                logger.info(`Unable to send email to ${tenant.Email}`);
            }
        }
        catch (error) {
            logger.info(`Unable to send email to ${tenant.Email}`);
        }
    };

    private rollbackCreateTenant = async (tenant: TenantDto) => {
        try {
            if (tenant) {
                await this._service.delete(tenant.id, true);
            }
        } catch (error) {
            logger.info(error);
        }
    };

}
